### PR TITLE
DT-896: Add custom configuration of jwt fields to clients

### PR DESCRIPTION
**Pull request by petergphillips**

### CODE REVIEW

File: build.gradle.kts [hunk 1]
```diff
@@ -8,21 +8,18 @@ repositories {
...
-  implementation("org.springframework.security:spring-security-jwt:1.1.0.RELEASE")
-  implementation("org.springframework.security.oauth:spring-security-oauth2:2.4.1.RELEASE")
+  implementation("org.springframework.security:spring-security-jwt:1.1.1.RELEASE")
+  implementation("org.springframework.security.oauth:spring-security-oauth2:2.5.0.RELEASE")
   implementation("io.jsonwebtoken:jjwt:0.9.1")
-  implementation("com.nimbusds:nimbus-jose-jwt:8.16")
+  implementation("com.nimbusds:nimbus-jose-jwt:8.19")
 
   implementation("org.springframework.boot:spring-boot-starter-data-jpa")
-  implementation("org.hibernate:hibernate-core:5.4.15.Final")
```
Comment (petergphillips) [inline build.gradle.kts:25]: included in data-jpa anyway

File: src/main/java/uk/gov/justice/digital/hmpps/oauth2server/config/TrackingTokenServices.kt [hunk 2]
```diff
@@ -74,7 +74,7 @@ open class TrackingTokenServices(private val telemetryClient: TelemetryClient,
   }
 
   private fun retrieveUsernameFromToken(token: OAuth2AccessToken): String {
-    val username = token.additionalInformation[JWTTokenEnhancer.ADD_INFO_USER_NAME] as String?
+    val username = token.additionalInformation[JWTTokenEnhancer.SUBJECT] as String?
```
Comment (petergphillips) [inline src/main/java/uk/gov/justice/digital/hmpps/oauth2server/config/TrackingTokenServices.kt:77]: subject will always be set and moving away from user name

File: src/main/resources/db/auth/V4_02__default_client_details_jwt_fields.sql [hunk 3]
```diff
@@ -0,0 +1,3 @@
+UPDATE oauth_client_details
+SET additional_information = '{"jwtFields":"-name,+user_name"}'
```
Comment (petergphillips) [inline src/main/resources/db/auth/V4_02__default_client_details_jwt_fields.sql:2]: set default to keep `user_name` and remove `name` for existing clients.

File: src/main/resources/db/dev/data/auth/V900_0__oauth_sample_data.sql [hunk 4]
```diff
@@ -1,5 +1,5 @@
 INSERT INTO oauth_client_details (client_id, access_token_validity, additional_information, authorities, authorized_grant_types, autoapprove, client_secret, refresh_token_validity, resource_ids, scope, web_server_redirect_uri)
-VALUES ('omicuser','1200',null,'SYSTEM_READ_ONLY','password,authorization_code,refresh_token','read','$2a$10$EiAoV/3ZSHl4KsVAkYZmH.pfZ6tLcK2vlvJTBgQBML3p3LrcPjaCi',null,null,'read','http://localhost:3000/login'),
+VALUES ('omicuser','1200','','SYSTEM_READ_ONLY','password,authorization_code,refresh_token','read','$2a$10$RYwV0QebHAovVXWPySb2lefr3HTDntGu1euXHDJc3zwh2NsqeNGHG',null,null,'read','http://localhost:3000/login,http://localhost:8081/login'),
```
Comment (petergphillips) [inline src/main/resources/db/dev/data/auth/V900_0__oauth_sample_data.sql:2]: use `''` as additional information so won't be set by the update lower down

File: src/test/java/uk/gov/justice/digital/hmpps/oauth2server/config/TrackingTokenServicesTest.kt [hunk 5]
```diff
@@ -71,7 +74,7 @@ internal class TrackingTokenServicesTest {
       val userAuthentication = UsernamePasswordAuthenticationToken(USER_DETAILS, "credentials")
       tokenServices.createAccessToken(OAuth2Authentication(OAUTH_2_REQUEST, userAuthentication))
       verify(restTemplate).postForLocation(eq("/token?authJwtId={authJwtId}"), check {
-        assertThat(it).isInstanceOf(String::class.java).asString().hasSize(28)
+        assertThat(it).isInstanceOf(String::class.java).asString().hasSize(27)
```
Comment (petergphillips) [inline src/test/java/uk/gov/justice/digital/hmpps/oauth2server/config/TrackingTokenServicesTest.kt:77]: change due to spring security oauth upgrade to 2.5.0

File: src/main/resources/templates/ui/form.html [hunk 6]
```diff
@@ -98,6 +98,14 @@ <h1 th:if="${param.client}" class="govuk-heading-l">Edit client '<span th:text="
       </div>
     </div>
 
+    <div class="govuk-form-group">
+      <div class="govuk-form-group">
+        <label class="govuk-label" for="clientId">JWT Fields Configuration</label>
+        <input class="govuk-input" id="jwtFields" type="text" name="jwtFields"
+               th:value="${clientDetails.additionalInformation['jwtFields']}"/>
```
Comment (Mjwillis) [inline src/main/resources/templates/ui/form.html:105]: Oooh nice!

File: src/test/resources/tokenverification/mappings/post_refresh_token.json [hunk 7]
```diff
@@ -1,7 +1,7 @@
 {
   "request": {
     "method": "POST",
-    "urlPattern": "/token/refresh\\?accessJwtId=([a-zA-Z0-9+%/]*)"
```
Comment (Mjwillis) [inline src/test/resources/tokenverification/mappings/post_refresh_token.json:4]: why this change?

Comment (petergphillips) [reply src/test/resources/tokenverification/mappings/post_refresh_token.json:4]: test was randomly failing as the mapping wasn't matching
